python/galois/shmem.pxd

```diff
@@ -0,0 +1,80 @@
+from cython.operator cimport preincrement, dereference as deref
+from libgalois.Galois cimport UserContext, iterate, for_each, setActiveThreads, SharedMemSys, loopname, no_conflicts, no_pushes, gPrint, do_all, GReduceMax, InsertBag, steal
+from libgalois.Galois cimport LargeArray, MethodFlag, FLAG_UNPROTECTED
+from libgalois.graphs.Graph cimport dummy_true, dummy_false, MorphGraph, LC_CSR_Graph
+from libgalois.graphs.Util cimport readGraph
+from libgalois.Timer cimport Timer
+from libstd.atomic cimport atomic
+from libcpp.vector cimport vector
+from libcpp.string cimport string
+from libcpp cimport bool
+import sys
+from libc.stdint cimport *
+from libc.math cimport fabs
+
+# Initialize the Galois runtime when the Python module is loaded.
+cdef class _galois_runtime_wrapper:
+    cdef SharedMemSys _galois_runtime
+
+cdef extern from * nogil:
+    # hack to bind leading arguments by value to something that can be passed
+    # to for_each. The returned lambda needs to be usable after the scope
+    # where it is created closes, so captured values are captured by value.
+    # The by-value capture in turn requires that graphs be passed as
+    # pointers. This function is used without exception specification under
+    # the assumption that it will always be used as a subexpression of
+    # a whole expression that requires exception handling or that it will
+    # be used in a context where C++ exceptions are appropriate.
+    # There are more robust ways to do this, but this didn't require
+    # users to find and include additional C++ headers specific to
+    # this interface.
+    # Syntactically, this is using the cname of an "external" function
+    # to create a one-line macro that can be used like a function.
+    # The expected use is bind_leading(function, args).
+    cdef void *bind_leading "[](auto f, auto&&... bound_args){return [=](auto&&... pars){return f(bound_args..., pars...);};}"(...)
+    # Similar thing to invoke a function and return an integer.
+    # Useful for verifying that this approach works.
+    cdef int invoke "[](auto f, auto&&... args){return f(args...);}"(...)
```

Comment:
We could potentially use std::invoke and std::bind here instead of these hacked-in lambdas. The real problem is that cython isn't capable of reasoning meaningfully about the return type of such a thing though.